Rewrite of iPXE chapter (#726)
theforeman/foreman-documentation: guides/doc-Provisioning_Guide/topics/proc_configuring_iPXE_to_reduce_provisioning_times.adoc

For more information, see https://ipxe.org/appnote/hardware_drivers[supported hardware on ipxe.org] for a list of hardware drivers expected to work with an iPXE-based boot disk.
* The NIC is compatible with iPXE.

To prepare iPXE environment, you must perform this procedure on all {SmartProxies}.

.Procedure
+
. Enable the *tftp* and *httpboot* services:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {foreman-installer} --foreman-proxy-httpboot true --foreman-proxy-tftp true
----

. Install the `ipxe-bootimgs` RPM package:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {package-install} ipxe-bootimgs
----
+
ifdef::foreman-el,katello[]
. On Debian/Ubuntu, install the `ipxe` .deb package:
+
----
# {package-install} ipxe
----
+
endif::[]
ifdef::foreman-el,katello[]
. On systems with SELinux, correct the file contexts:
+
----
# restorecon -RvF /var/lib/tftpboot/
----
endif::[]
ifdef::satellite,orcharhino[]

. Correct the SELinux file contexts:
+
----
# restorecon -RvF /var/lib/tftpboot/
----
endif::[]

> * Copy the iPXE firmware with Linux kernel header to the TFTP directory:

Reviewer: ```suggestion
* Copy the iPXE firmware with the Linux kernel header to the TFTP directory:
```

For more information, see https://ipxe.org/appnote/hardware_drivers[supported hardware on ipxe.org] for a list of hardware drivers expected to work with an iPXE-based boot disk.
* The NIC is compatible with iPXE.

.Procedure

. To prepare iPXE environment, you must perform this procedure on all {SmartProxies}.
+
. Enable the *tftp* and *httpboot* services:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {foreman-installer} --foreman-proxy-httpboot true --foreman-proxy-tftp true
----
+
. Install the `ipxe-bootimgs` RPM package:
+
[options="nowrap" subs="+quotes,attributes"]
----
# {package-install} ipxe-bootimgs
----
+
ifdef::foreman-el,katello[]
. On Debian/Ubuntu, install the `ipxe` .deb package:
+
----
# {package-install} ipxe
----
+
endif::[]
ifdef::foreman-el,katello[]
. On systems with SELinux, correct the file contexts:
+
----
# restorecon -RvF /var/lib/tftpboot/
----
endif::[]
ifdef::satellite,orcharhino[]
. Correct the SELinux file contexts:
+
----
# restorecon -RvF /var/lib/tftpboot/
----
endif::[]
+
* Copy the iPXE firmware with Linux kernel header to the TFTP directory:
+
----
# cp /usr/share/ipxe/ipxe.lkrn /var/lib/tftpboot/
----
+

> * Copy the UNDI iPXE firmware to the TFTP directory:

Author: Explained.

+
----
# cp /usr/share/ipxe/undionly.kpxe /var/lib/tftpboot/undionly.0
----
+
. Optionally, configure Foreman discovery:
+
* Install Foreman discovery image:
+
[options="nowrap" subs="+quotes,attributes"]

> ifdef::satellite[]

Reviewer 2: Is this ifdef=satellite really necessary? I think this also applies to Foreman and orcharhino.

Author: There is no RPM package upstream, let me do a xref instead.

. Virtual machine starts
. iPXE retrieves the network credentials using DHCP
. iPXE retrieves the HTTP address using DHCP

> . iPXE loads the iPXE bootsrap template from the template {SmartProxy}

Reviewer 2: ```suggestion
. iPXE loads the iPXE bootstrap template from {SmartProxy}
```

> . iPXE loads the iPXE template with MAC as a URL parameter from the template {SmartProxy}

Reviewer 2: ```suggestion
. iPXE loads the iPXE template with MAC as a URL parameter from {SmartProxy}
```

.Procedure

> 

Reviewer 2: ```suggestion
```

. In the *Hosts* page, select the host that you want to use.
. Select the *Templates* tab.
. From the *iPXE template* list, select *Review* to verify that the *Kickstart default iPXE* template is the correct template.
. Configure the `dhcpd.conf` file as follows:
+
[options="nowrap" subs="+quotes,attributes"]
----
if exists user-class and option user-class = "iPXE" {
  filename "http://{smartproxy-example-com}:8000/unattended/iPXE?bootstrap=1";
}
----
+
If you use an isolated network, use a {SmartProxyServer} URL with TCP port `8000`, instead of the URL of {ProjectServer}.
+
[NOTE]

> Use a different port than `8000` when changed with `--foreman-proxy-http-port installer` option.

Reviewer 2: ```suggestion
If you have changed the port using the `--foreman-proxy-http-port installer` option, use your custom port.
```

. iPXE chainloads the iPXE template from the template {SmartProxy}
. iPXE loads the kernel and initial RAM disk of the installer

.Configuring {ProjectName} Server to use iPXE

You can use the default template to configure iPXE booting for hosts.
If you want to change the default values in the template, clone the template and edit the clone.

> . In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates*, enter `PXELinux chain iPXE` to use `ipxe.lkrn` or, for BIOS systems, enter `PXELinux chain iPXE UNDI` to use `undionly-ipxe.0`, and click *Search*.

Reviewer 2: ```suggestion
. In the {ProjectWebUI}, navigate to *Hosts* > *Provisioning Templates*.
. Enter `PXELinux chain iPXE` to use `ipxe.lkrn` or, for BIOS systems, enter `PXELinux chain iPXE UNDI` to use `undionly-ipxe.0`, and click *Search*.
```

. From the *iPXE template* list, select *Review* to verify the template is the correct template.
If there is no PXELinux entry, or you cannot find the new template, navigate to *Hosts* > *All Hosts*, and on your host, click *Edit*.
Click the *Operating system* tab and click the Provisioning Template *Resolve* button to refresh the list of templates.
. Configure the `dhcpd.conf` file as follows:
+
[options="nowrap" subs="+quotes,attributes"]
----
if exists user-class and option user-class = "iPXE" {
  filename "http://{smartproxy-example-com}:8000/unattended/iPXE?bootstrap=1";
} else {
  filename "undionly.0";
}
----
+
[NOTE]

> Use a different port than `8000` when changed with `--foreman-proxy-http-port installer` option.

Reviewer 2: See above.
